spec/e2e/unpersisted_association_spec.rb

```diff
@@ -0,0 +1,180 @@
+        end
+
+        it 'saves both nodes and creates the relationship' do
+          expect(math).to receive(:save).and_call_original
+          expect { chris.save }.to change { chris.favorite_class }
+        end
+      end
+    end
+  end
+
+  describe 'has_many' do
+    context 'with persisted nodes' do
+      let(:chris) { Student.create(name: 'Chris') }
+      let(:math)  { Lesson.create(subject: 'Math') }
+
+      it 'creates a relationship' do
+        expect { chris.lessons << math }.to change { chris.lessons.count }
+      end
+    end
+
+    context 'between two unpersisted nodes' do
+      let(:chris) { Student.new(name: 'Chris') }
+      let(:math)  { Lesson.new(subject: 'Math') }
+
+      it 'does not raise an error' do
+        expect { chris.lessons << math }.not_to raise_error
+      end
+
+      it 'does not create new nodes' do
+        expect { chris.lessons << math }.not_to change { Student.count == 0 && Lesson.count == 0 }
+      end
+
+      it 'is aware that there are cascading relationships' do
+        expect { chris.lessons << math }.to change { chris.pending_associations? }
+      end
+
+      context 'upon save...' do
+        before { chris.lessons << math }
+
+        it 'saves both nodes' do
+          expect { chris.save }.to change { math.exist? }
+        end
+
+        it 'creates the relationship' do
+          save_and_expect_rel!.call
+        end
+
+        context 'with a save failure' do
+          let(:unnamed_lesson) { Lesson.new }
+          before do
+            expect(unnamed_lesson).not_to be_valid
+            chris.lessons << unnamed_lesson
+          end
+
+          it 'rolls back the entire transaction' do
+            expect { chris.save }.to raise_error
+            expect(chris).not_to exist
+          end
+        end
+      end
+    end
+
+    context 'between many unpersisted nodes' do
+      let(:chris) { Student.new(name: 'Chris') }
+      let(:math)  { Lesson.new(subject: 'Math') }
+      let(:science) { Lesson.new(subject: 'Science') }
+      let(:lessons) { [math, science] }
+
+      context 'associated as an array' do
+        it 'delays the call to :save' do
+          expect(science).not_to receive(:save)
+          expect { chris.lessons << lessons }.not_to raise_error
+        end
+
+        it 'calls save on each element' do
+          expect(math).to receive(:save).and_call_original
+          expect(science).to receive(:save).and_call_original
+          chris.lessons << lessons
+          chris.save
+        end
+      end
+
+      context 'associated individually' do
+        it 'calls save on each node' do
+          expect(math).to receive(:save).and_call_original
+          expect(science).to receive(:save).and_call_original
+          chris.lessons << math
+          chris.lessons << science
+          chris.save
+        end
+      end
+    end
+
+    context 'between unpersisted and persisted, unchanged nodes' do
+      let(:chris) { Student.new(name: 'Chris') }
```

Comment:
I've been feeling like the better approach a lot of times for tests is to create variable names like `persisted_student`/`unpersisted_student`/etc... because it makes it easier to tell what's going on.  This also means that you could move the `let` definitions to the top of the main block because since `let` is lazily loaded these objects would only get made / persisted when used.

Sometimes when you move `let` variables to the top like that you get problems forgetting which variables do what, but I think if you name according to their purpose for the test (like above) then that's not a problem anymore.